### PR TITLE
MGDOBR-647: Refactoring kustomize project

### PR DESCRIPTION
**Pull request by r00ta**
[MGDOBR-647](https://issues.redhat.com/browse/MGDOBR-647)

The idea of this refactoring is the following: 

- We still keep a `base` layer, which is common for openshift and k8s
- We still keep a `minikube` overlay that uses `base`
- We add a `base-openshift` layer that contains all the configurations in common for `dev` and `stable` environments. This layer uses `base`
- We keep the overlays `dev` and `stable`, but this time they contain only the specific configurations for the environments (for example, secrets). 

The new deployment strategy is the following: 
- ArgoCD on `dev` env will watch this repository under the `dev` branch. 
- ArgoCD on `stable` env will watch this repository under the `stable` branch

When you want to deploy something to `dev`, just open a PR from `main` to `dev` and merge it. 
When you want to deploy something to `stable`, just open a PR from `dev` to `stable` and merge it. 

Note that this time if you are making changes to the code and an update of the kustomization files is needed, you have to do it in the PR itself for both `dev` and `stable` overlays. 
Since the actual deployment of the resources is done from `dev` and `stable` branch, on `main` the code and the kustomize layer must be aligned. 

For the time being, we still need the bot that updates the images on `main` when they are available. 
 
## PR owner checklist

Please make sure that your PR meets the following requirements:

- [ ] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [ ] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [ ] All new functionality is tested
- [ ] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.

</details>

### CODE REVIEW

**Review by danielezonca — APPROVED** [review]
LGTM
Make sure to update the guide to promote from dev to stable and [the scripts](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/dev/bin/create-stable-pr.sh)

File: kustomize/overlays/stable/secrets/event-bridge-manager-secrets.yaml [hunk 1]
```diff
@@ -32,6 +32,9 @@ spec:
...
     SHARD_ID: AgCQiLbXmBtRGK0BOnh3Vjhq5HHJGU16UmDE8C9+wHq1vU2iTOXx03blrQi3CPFAQ2z4idU8BwzWYNXZZmatuxANXc8wW5o6V7GZ7qgVnFUzz9oYXA3vilBDw1OXOWdNRnG/0n/quf6TB1ImlyJBw+ZRAxFrIuZObmC9xEpIv5uX1JSzjAJBBrGnsPnBegrbIhUkN103DbmDQWx2RCMtafLqRe/79NTrXJ/RO/zO2QLKGQH91ePq6b1t8giwCq+Q74EhfZ/1Ga2xLj+8GzbVSUQpxgE8YQxnxI/8LqQzwkhfkJmUtOeq2H/NxAdmEJNigYAn7SfmdzyNf8BugX9E74EEJbR7WLjwma4/htOzQpZfjUAvUuBDU19u+7XGDqv0awsrDUW0kPG4Ui7RpnzmvKeqG5uiW9pMBf90v5tbG25OXC+BNtjACZm9LhKQT1Zo4mtKwyUa5KrcLAwdVeh1v9oGXJwBRhkKLNHdJ+xJSB0bREwwH/j8H/axCSQ5Uz4oQ4Pd62+rbXSDZtffsC8Et7TNbiOEAG0WP1ug/gOCf9HXWyK//Vbvfou0CnPDF3i3nP27+6CFrxgMPNAc5UiB2I7xVXbXA+azEOod929QTGFG1L900yr3WbLZnojydKxqtxmDhYEDmar64Fsq8JMC6eYMq5PEdlqsOpHgj35edrinKyOOuRfmM/ghZmr1J7GGdhpjqti32xjyIg==
+    # TODO: remove the following property after the system is updated to use "event-bridge.resource-prefix" property
```
[the remark this answers is not in the document]
Comment (kostola) [reply kustomize/overlays/stable/secrets/event-bridge-manager-secrets.yaml:35]: I think this line and the `EVENT_BRIDGE_KAFKA_TOPIC_PREFIX` secret can be removed now

File: kustomize/overlays/dev/secrets/event-bridge-manager-secrets.yaml [hunk 2]
```diff
@@ -32,6 +32,9 @@ spec:
...
     SHARD_ID: AgBk9cQqlSbuZlRbPqpq6epMdoF6uIpaihGhHuMAtujx6rDnQIIVQUO65L7UmuTrpnwrgunCRXrJAWiXRS2UJgMhSgkB1E4QyYSWtR0MwXYYqQwBefFeoP9iToL0k3PCx8CMzTunNtvj2AEAf7jPw0X77OIq8BWHYtQLB1p8xszu5dbl2VwxUIAZobL4lxspjTkuFQWhXuEIUao227IsZJ684P8eYzuSM0wMsw0LdsJts0azBuFJiZGwzN8hOxGeluPKxpa5srQwnjB7EGTnhowigMIBBiJYMULcXSYGTnRvC2GazLjjEZeuDczQfgBiYPAgj0TClTR6xqIUqCJgVw9jwWlo3zABCo2G6LqkIADaz395Qoix3aTfc+9h6VhIQg/17InYdqLdRiV1Xg+YVH8RIITYOAA1He1DHEFZADB1IMHyZmorrX05+0FHJao1QKvdq6t6oPr9dt/048Kf9ELqi1+yUBPWwIX2C5UfRmSOYnBiud2DqR0BwwaoHq+1iHo6YO4nF7ChDYLAfN6g8mOuHajB5vz0aGnnZ1eP/JCypOXU744QVbB8+2SbwALX0f8qlmXy8Bp0ko5gLRrB69GtIi1QGCGQccEWyBFgd6DQ3k/cYxWAujFceooQpeeQ3/ijoPhjOG5J1vSKGp/MrOASwbpnHxGbyWkrThxYNusYNSFM/uKE/s76pFtuYkTyRC/vYabE5y9qiQ==
+    # TODO: remove the following property after the system is updated to use "event-bridge.resource-prefix" property
```
[the remark this answers is not in the document]
Comment (kostola) [reply kustomize/overlays/dev/secrets/event-bridge-manager-secrets.yaml:35]: Same [here](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/689/files/d353aa4cc8a1ecc19c625c2e9b8f93d1bfbd95f6#r860901686)

**Review by robpblake — COMMENTED** [review]
@r00ta I'm +1 on your proposed layout for the kustomize directory, but I really don't see the need for a `dev` and `stable` branch. The only reason I can see to do that would be to model environment specific configuration and that's exactly what `kustomize` does for us with the overlays.

As you say yourself in the PR description: environment specific configuration goes into the overlay for the environment. What else does the branch give us over that?

Part of what started this discussion was the ability to have different versions of the CRDs for the Fleet Shard operator in each environment when we make changes (as we have done in with @kostola work). Why don't we just have the CRD version that matches the operator in the overlay for the environment? So that would mean something like:

- kustomize/overlays/dev/shard/resources/bridgeexecutors.com.redhat.service.bridge-v1.yml
- kustomize/overlays/stable/shard/resources/bridgeexecutors.com.redhat.service.bridge-v1.yml

Where the CRD matches the version of the Operator deployed in that environment, rather than the CRD being defined in `base`.


cheers,

Rob

**Comment by r00ta** [comment]
> @r00ta I'm +1 on your proposed layout for the kustomize directory, but I really don't see the need for a `dev` and `stable` branch. The only reason I can see to do that would be to model environment specific configuration and that's exactly what `kustomize` does for us with the overlays.
> 
> As you say yourself in the PR description: environment specific configuration goes into the overlay for the environment. What else does the branch give us over that?
> 
> Part of what started this discussion was the ability to have different versions of the CRDs for the Fleet Shard operator in each environment when we make changes (as we have done in with @kostola work). Why don't we just have the CRD version that matches the operator in the overlay for the environment? So that would mean something like:
> 
> * kustomize/overlays/dev/shard/resources/bridgeexecutors.com.redhat.service.bridge-v1.yml
> * kustomize/overlays/stable/shard/resources/bridgeexecutors.com.redhat.service.bridge-v1.yml
> 
> Where the CRD matches the version of the Operator deployed in that environment, rather than the CRD being defined in `base`.
> 
> cheers,
> 
> Rob

Let's consider the following scenario: we do merge some changes and a new configuration `MY_CONFIG` is needed for the manager. 
- With the current implementation we have to: 
  - add the `MY_CONFIG` configuration in the `dev` overlay
  - update the images 
  - When we are happy about the changes on `dev`, we have to copy/paste the new config from `dev` to the `stable` overlay together with the new images. 
- With the new strategy we have to: 
  - add the `MY_CONFIG` configuration in the `base-openshift` overlay
  - add the new images in the `base-openshift` overlay
  - Create a PR from `main` to `dev` (it will just perform a fast-forward) -> the changes are deployed to `dev`
  - Once we are happy with the changes on `dev`, we open a PR from `dev` to `stable` (it will just perform a fast-forward) -> the changes are deployed to `stable`. 
  
Which of the two approches do you think is more error prone?  Consider the case in which `stable` is behind ~10 versions respect to `dev`: it's really hard to remember/time consuming to remember if there are properties to be copied from `dev` overlay to `stable`.

**Comment by r00ta** [comment]
> Where the CRD matches the version of the Operator deployed in that environment, rather than the CRD being defined in `base`.
> 
> cheers,
> 
> Rob

Maybe there was a misunderstanding: with this PR I'm trying to improve the entire situation, not only the problem with the CRDs we talked about few days ago.

**Comment by sutaakar** [comment]
This PR is blocking basically any E2E test from PR check (due to missing CRD adjustments for Processor type). Please merge it soon.

File: kustomize/overlays/stable/secrets/event-bridge-manager-secrets.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (r00ta) [reply kustomize/overlays/stable/secrets/event-bridge-manager-secrets.yaml]: Done, thanks :+1:

**Comment by r00ta** [comment]
> LGTM Make sure to update the guide to promote from dev to stable and [the scripts](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/dev/bin/create-stable-pr.sh)

I removed the script (cc @lucamolteni ) as we don't need anymore to copy any configuration from `dev` to `stable`

**Pull request merged**